src/Diagnostics.RuntimeHost/Services/SourceWatcher/Watchers/StorageWatcher.cs

```diff
@@ -218,5 +251,47 @@ private async Task StartBlobDownload(bool startup = false)
             
         }
 
+        private async Task StartKustoCacheRefresh(bool startup = false)
+        {
+            var diagConfigRows = await storageService.GetKustoConfiguration();
+            var configsToLoad = new List<DiagConfiguration>();
+            if(!startup)
+            {
+                configsToLoad.AddRange(diagConfigRows.Where(row => row.Timestamp >= kustoCacheLastModifiedTime).ToList());
+            } else
+            {
+                configsToLoad.AddRange(diagConfigRows);
+            }
+            var stopwatch = new Stopwatch();
+            stopwatch.Start();
+            try
+            {
+                if (configsToLoad.Count > 0)
+                {
+                    DiagnosticsETWProvider.Instance.LogAzureStorageMessage(nameof(StorageWatcher), $"Starting kusto config download to update cache, Number of configs to load: {configsToLoad.Count}, startup : {startup.ToString()} at {DateTime.UtcNow}");
+                }
+                foreach(var config in configsToLoad)
+                {
+                    var kustoMappingsStringContent = config.KustoClusterMapping;
+                    var kustoMappings = (List<Dictionary<string, string>>)JsonConvert.DeserializeObject(kustoMappingsStringContent, typeof(List<Dictionary<string, string>>));
+                    var resourceProvider = config.RowKey;
+                    if (!kustoCacheService.ContainsKey(resourceProvider) || (kustoCacheService.TryGetValue(resourceProvider, out List<Dictionary<string, string>> value) && !value.Equals(kustoMappings)))
+                    {
+                        kustoCacheService.AddOrUpdate(resourceProvider, kustoMappings);
+                    }
+                }
```

Comment:
Should this `foreach` be within the aforementioned if clause `if (configsToLoad.Count > 0)`

Reply:
Technically it doesn't matter because foreach of an empty list yields nothing. So either way I'm fine with it 👍🏿